launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java

```diff
@@ -0,0 +1,266 @@
+    res.getWriter().println(formHtml);
+  }
+
+  /**
+   * Generate policy for directly uploading a file to Cloud Storage via HTML form.
+   * @return HTML form (as a String) for uploading MP3
+   */
+  public String generateSignedPostPolicyV4 (String projectId, String bucketName, String blobName) throws IOException {
+    // String blobName = entity ID from Datastore, as the name of object uploaded to GCS
+
+    String errorMessage = "Policy is null.";
+    Storage storage;
+    PostPolicyV4.PostFieldsV4 fields;
+    PostPolicyV4 policy = null;
+    String myRedirectUrl = "";
+    try {
+      // Prepare credentials and API key
+      String keyFileName = "launchpod-step18-2020-47434aafba88.json";
+      ClassLoader classLoader = getClass().getClassLoader();
+      File file = new File(classLoader.getResource(keyFileName).getFile());
+      GoogleCredentials credentials = GoogleCredentials.fromStream(new FileInputStream(file));
+
+      storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
+
+      myRedirectUrl="https://launchpod-step18-2020.appspot.com/rss-feed?action=generateRSSLink&id=" + blobName;
+      fields =
+          PostPolicyV4.PostFieldsV4.newBuilder().setSuccessActionRedirect(myRedirectUrl).build();
+
+      policy =
+        storage.generateSignedPostPolicyV4(
+            BlobInfo.newBuilder(bucketName, blobName).build(), 10, TimeUnit.MINUTES, fields);
+    } catch(Exception e) {
+      errorMessage = e.getMessage();
+      throw e;
+    }
+
+    StringBuilder htmlForm;
+    if (policy!=null) {
+      htmlForm =
+          new StringBuilder(
+              "<form name='mp3-upload' action='"
+                  + policy.getUrl()
+                  + "' method='POST' enctype='multipart/form-data'>\n");
+      for (Map.Entry<String, String> entry : policy.getFields().entrySet()) {
+        htmlForm.append(
+            "  <input name='"
+                + entry.getKey()
+                + "' value='"
+                + entry.getValue()
+                + "' type='hidden' />\n");
+      }
+      htmlForm.append("  <input type='file' name='file'/><br />\n");
+      htmlForm.append("  <input type='submit' value='Upload File' name='submit'/><br />\n");
+      htmlForm.append("</form>\n");
+    } else {
+      htmlForm = new StringBuilder();
+      htmlForm.append(errorMessage);
+    }
+    return htmlForm.toString();
+  }
+
+  /**
+  * Display RSS feed xml string that user tries recalling with the given ID
+  */
+  @Override
+  public void doGet(HttpServletRequest req, HttpServletResponse res) throws IOException {
+    String action = req.getParameter(ACTION);
+    String id = req.getParameter(ID);
+
+    if (action==null || id==null) {
+      res.setContentType("text/html");
+      res.getWriter().println("Please specify action and/or id.");
+      res.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    if (action.equals("generateRSSLink")) {
```

Comment:
it might look nicer to turn action into a enum. you can convert a string to a enum with valueOf()

Reply:
valueOf() only takes in a String that is an exact match for the enum constant itself. I changed the code to use enums, but I'm not sure if this looks any better haha